oracle/oracles/pd.go

```diff
@@ -366,3 +610,81 @@ func (o *pdOracle) SetExternalTimestamp(ctx context.Context, ts uint64) error {
 func (o *pdOracle) GetExternalTimestamp(ctx context.Context) (uint64, error) {
 	return o.c.GetExternalTimestamp(ctx)
 }
+
+func (o *pdOracle) getCurrentTSForValidation(ctx context.Context, opt *oracle.Option) (uint64, error) {
+	ch := o.tsForValidation.DoChan(opt.TxnScope, func() (interface{}, error) {
+		//metrics.ValidateReadTSFromPDCount.Inc()
+
+		// If the call that triggers the execution of this function is canceled by the context, other calls that are
+		// waiting for reusing the same result should not be canceled. So pass context.Background() instead of the
+		// current ctx.
+		res, err := o.GetTimestamp(context.Background(), opt)
+		// After finishing the current call, allow the next call to trigger fetching a new TS.
+		o.tsForValidation.Forget(opt.TxnScope)
+		return res, err
+	})
+	select {
+	case <-ctx.Done():
+		return 0, errors.WithStack(ctx.Err())
+	case res := <-ch:
+		if res.Err != nil {
+			return 0, errors.WithStack(res.Err)
+		}
+		return res.Val.(uint64), nil
+	}
+}
+
+func (o *pdOracle) ValidateSnapshotReadTS(ctx context.Context, readTS uint64, opt *oracle.Option) error {
+	latestTS, err := o.GetLowResolutionTimestamp(ctx, opt)
+	// If we fail to get latestTS or the readTS exceeds it, get a timestamp from PD to double-check.
+	// But we don't need to strictly fetch the latest TS. So if there are already concurrent calls to this function
+	// loading the latest TS, we can just reuse the same result to avoid too many concurrent GetTS calls.
+	if err != nil || readTS > latestTS {
+		currentTS, err := o.getCurrentTSForValidation(ctx, opt)
+		if err != nil {
+			return errors.Errorf("fail to validate read timestamp: %v", err)
+		}
+		o.adjustUpdateLowResolutionTSIntervalWithRequestedStaleness(readTS, currentTS, time.Now())
+		if readTS > currentTS {
+			return errors.Errorf("cannot set read timestamp to a future time")
+		}
+	} else {
+		estimatedCurrentTS, err := o.getStaleTimestamp(opt.TxnScope, 0)
+		if err != nil {
+			logutil.Logger(ctx).Warn("failed to estimate current ts by getSlateTimestamp for auto-adjusting update low resolution ts interval",
+				zap.Error(err), zap.Uint64("readTS", readTS), zap.String("txnScope", opt.TxnScope))
+		} else {
+			o.adjustUpdateLowResolutionTSIntervalWithRequestedStaleness(readTS, estimatedCurrentTS, time.Now())
+		}
```

Comment:
I'm sorry, I don't understand why we need those logic?

Reply:
We estimate the gap between the current time and the readTS, to see if the gap is shorter than the update interval, so that we need to shrink the update interval.